### PR TITLE
[Data] Add `partition_filter` parameter to `read_parquet`

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

If your dataset contains any non-parquet files, `read_parquet` errors. To workaround this, this PR adds a `partition_filter` parameter so users can ignore non-parquet files.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/parquet_datasource.py [hunk 1]
```diff
@@ -199,6 +196,22 @@ def __init__(
...
             )
 
+        paths, filesystem = _resolve_paths_and_filesystem(paths, filesystem)
+
+        # HACK: PyArrow's `ParquetDataset` errors if input paths contain non-parquet
+        # files. To avoid this, we expand the input paths with the default metadata
+        # provider and then apply the partition filter.
+        partition_filter = reader_args.pop("partition_filter", None)
+        if partition_filter is not None:
+            default_meta_provider = DefaultFileMetadataProvider()
```
Comment (c21) [inline python/ray/data/datasource/parquet_datasource.py:206]: should we call `get_generic_metadata_provider()` instead?

File: python/ray/data/tests/test_parquet.py [hunk 2]
```diff
@@ -756,6 +757,20 @@ def test_parquet_write(ray_start_regular_shared, fs, data_path, endpoint_url):
...
+    pq.write_table(empty_table, tmp_path / "table.parquet")
+    # `spam` should be filtered out.
+    with open(tmp_path / "spam", "w"):
+        pass
+
+    ds = ray.data.read_parquet(
+        tmp_path, partition_filter=FileExtensionFilter("parquet")
+    )
+
+    assert ds.count() == 0
```
Comment (c21) [inline python/ray/data/tests/test_parquet.py:771]: sorry if I am wrong, but shouldn't `table.parquet` still there?

File: python/ray/data/datasource/parquet_datasource.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/data/datasource/parquet_datasource.py:206]: Ah, yeah, good call

File: python/ray/data/tests/test_parquet.py (hunk 2 above)
Comment (c21) [reply python/ray/data/tests/test_parquet.py:771]: ok, how about creating a non-empty table, so we can test out the filtering works, and not blindly drops all paths?

**Review by c21 — APPROVED** [review]
LG. let's test with user before merging, thanks.

**Pull request merged**